ktor-client/ktor-client-ios/darwin/src/io/ktor/client/engine/ios/IosClientEngine.kt

```diff
@@ -4,154 +4,74 @@
+            content.toNSData()?.let {
+                setHTTPBody(it)
+            }
 
             mergeHeaders(data.headers, data.body) { key, value ->
-                nativeRequest.setValue(value, key)
+                setValue(value, key)
             }
 
-            nativeRequest.setCachePolicy(NSURLRequestReloadIgnoringCacheData)
-            nativeRequest.setHTTPMethod(data.method.value)
-
-            launch(callContext) {
-                val content = data.body
-                val body = when (content) {
-                    is OutgoingContent.ByteArrayContent -> content.bytes().toNSData()
-                    is OutgoingContent.WriteChannelContent -> writer(dispatcher) {
-                        content.writeTo(channel)
-                    }.channel.readRemaining().readBytes().toNSData()
-                    is OutgoingContent.ReadChannelContent -> content.readFrom().readRemaining().readBytes().toNSData()
-                    is OutgoingContent.NoContent -> null
-                    else -> throw UnsupportedContentTypeException(content)
-                }
-
-                body?.let { nativeRequest.setHTTPBody(it) }
-
-                config.requestConfig(nativeRequest)
-                val task = session.dataTaskWithRequest(nativeRequest)
-                continuation.invokeOnCancellation { cause ->
-                    if (cause != null && task.state == NSURLSessionTaskStateRunning) {
-                        task.cancel()
-                    }
-                }
-                task.resume()
-            }.invokeOnCompletion {
-                session.finishTasksAndInvalidate()
-            }
+            setCachePolicy(NSURLRequestReloadIgnoringCacheData)
+            setHTTPMethod(data.method.value)
+
+            config.requestConfig(this)
         }
-    }
-}
 
-/**
- * Update [NSMutableURLRequest] and setup timeout interval that equal to socket interval specified by [HttpTimeout].
- */
-private fun NSMutableURLRequest.setupSocketTimeout(requestData: HttpRequestData) {
-    // iOS timeout works like a socket timeout.
-    requestData.getCapabilityOrNull(HttpTimeout)?.socketTimeoutMillis?.let {
-        if (it != HttpTimeout.INFINITE_TIMEOUT_MS) {
-            // Timeout should be specified in seconds.
-            setTimeoutInterval(it / 1000.0)
-        } else {
-            setTimeoutInterval(Double.MAX_VALUE)
+        val session = NSURLSession.sessionWithConfiguration(
+            configuration, responseReader, delegateQueue = NSOperationQueue.mainQueue()
+        )
+
+        val task = session.dataTaskWithRequest(nativeRequest)
+
+        launch(callContext) {
+            task.resume()
+        }
+
+        callContext[Job]!!.invokeOnCompletion {
+            session.finishTasksAndInvalidate()
+        }
+
+        return try {
+            responseReader.awaitResponse()
+        } catch (cause: CancellationException) {
+            if (task.state == NSURLSessionTaskStateRunning) {
+                task.cancel()
```

Comment:
Don't we have a race here between `task.resume()` inside of the launching coroutine and cancellation here? Is it legal to invoke `resume()` after `cancel()`?